hoomd/md/NeighborList.cc

```diff
@@ -1742,6 +1764,253 @@ void NeighborList::updateMemoryMapping()
+            }
+        }
+
+    unsigned long shape[] = {pair_list->size(), 2};
+    unsigned long stride[] = {2 * sizeof(uint32_t), sizeof(uint32_t)};
+
+    return pybind11::array_t(shape,                        // shape
+                             stride,                       // stride
+                             (uint32_t*)pair_list->data(), // data pointer
+                             delete_when_done);
+    }
+
+pybind11::object NeighborList::getPairListPython(uint64_t timestep)
+    {
+    compute(timestep);
+
+    bool third_law = getStorageMode() == NeighborList::half;
+
+#ifdef ENABLE_MPI
+    // if we are not the root processor, return None
+    bool root = m_exec_conf->isRoot();
+#endif
+
+    ArrayHandle<unsigned int> h_n_neigh(getNNeighArray(), access_location::host, access_mode::read);
+    ArrayHandle<unsigned int> h_nlist(getNListArray(), access_location::host, access_mode::read);
+    ArrayHandle<size_t> h_head_list(getHeadList(), access_location::host, access_mode::read);
+
+    ArrayHandle<unsigned int> h_tags(m_pdata->getTags(), access_location::host, access_mode::read);
+
+    auto* pair_list = new std::vector<vec2<uint32_t>>();
+
+    for (int i = 0; i < (int)m_pdata->getN(); i++)
+        {
+        unsigned int tag_i = h_tags.data[i];
+        const size_t myHead = h_head_list.data[i];
+        const unsigned int size = h_n_neigh.data[i];
+        for (unsigned int k = 0; k < size; k++)
+            {
+            const unsigned int j = h_nlist.data[myHead + k];
+            const unsigned int tag_j = h_tags.data[j];
+            if ((!third_law || j >= m_pdata->getN()) && tag_i > tag_j)
+                continue;
+
+            pair_list->push_back(vec2(tag_i, tag_j));
+            }
+        }
+
+    // pair_list->shrink_to_fit();
+
+#ifdef ENABLE_MPI
+
+    uint32_t* global_pair_list;
+    pybind11::capsule free_when_done;
+    unsigned long shape[] = {0, 2}; // first index needs to be set later
+    unsigned long stride[] = {2 * sizeof(uint32_t), sizeof(uint32_t)};
+
+    if (m_sysdef->isDomainDecomposed())
+        {
+        auto n_elem = 2 * (int)pair_list->size();
+        std::vector<int>* global_n_elem;
+        void* recvbuf = nullptr;
+        if (root)
+            {
+            global_n_elem = new std::vector<int>();
+            for (unsigned int i = 0; i < m_exec_conf->getNRanks(); i++)
+                {
+                global_n_elem->push_back(0);
+                }
+            recvbuf = global_n_elem->data();
+            }
+
+        // Use a gather command to get the number of pairs on each processor
+        MPI_Gather(&n_elem, 1, MPI_INT, recvbuf, 1, MPI_INT, 0, m_exec_conf->getMPICommunicator());
+
+        // Allocate array to hold the offsets
+        std::vector<int>* offsets;
+        int* recvcount = nullptr;
+        int* displs = nullptr;
+        int total_elements = 0;
+        if (root)
+            {
+            offsets = new std::vector<int>();
```

Comment:
```suggestion
            offsets = new std::vector<size_t>(https://en.cppreference.com/w/cpp/algorithm/partial_sum, 0);
```
No need to reserve and push back, just initialize to the correct size.